### ISSUE TITLE
Transcoding issues - ideas

### ISSUE THREAD

**Issue by openhs** [ampache/ampache#1048]
When transcoding is used several issues pops up.  Most of them have been reported (#135, #335, #522) and although they are closed they still persists in reality.  For that reason Ampache usability is quite limited with transcoding enabled.

From what I understood the core of the problem is that size/length of a track is not known upfront and the data are streamed as they are being transcoded.  I was thinking if it would be possible to transcode the track upfront and then send it to client as a normal file?  Personally I would rather wait until the track is fully transcoded rather than having the issues mentioned above.  I am not sure if it is technically possible but to have it at least as a configuration option would be splendid.  What do you think?

Moreover additional improvements could be build on top of that aiming to mitigate the pause during transcoding.

1. Server could implement caching.  Transcoded tracks would be saved and when requested again the saved file would be served without the delay.  Cache could be size-limited.
2. If client knows the next track it could request it upfront.  With server caching enabled the client does not even need to process the result (it can simply ignore or throw away the track that server sends as a result of the upfront request).  The transcoded (next) track would be cached on the server and when the client would want to play it for real it would be served without the delay.
3. Cache could be optimized to store only file metadata, like file size, file name, track length,...  This way the server would know how the result of transcoding will look like and send a "fake" file to the client without a delay while serving the content of the file as it is being transcoded.

I'm not sure what of those is actually technically possible with a web server though.  Hoping for the best. ;-)

**Comment by Afterster** [comment]
I agree with you but this would have a lot of stuff server-side, not sure you will see this feature before a long time as there is already lot of other feature requests. Any help appreciated guys :).

**Comment by Psy-Virus** [comment]
Maybe it's not that clean but why is Ampache not simply getting the source file length? Yeah okay, we don't know that the transcoding process works correctly, but as I see it we can trust in the work of ffmpeg/avconv (stable) I never even had any issues with these programs if I used healthy files. I also never saw differences in length, except the transcoding process was interrupted obviously.

Coding this with an option in the config file (to turn it on or off) and a warning that this COULD make problems is maybe the best way to get it working without this length issue until anyone finds a better way.

I know everyone prefers the clean and perfect way but then the process of transcoding wouldn't be needed because everyone would transcode their whole music library offline into the right format so Ampache could stream them directly.

My two cents.

:-)
